### ISSUE TITLE
ZFS receive dataset busy - race with /lib/udev/zvol_id

### ISSUE THREAD

**Issue by zviratko** [zfsonlinux/zfs#6504]
### System information
Type                                | Version/Name
  ---                                  |     --- 
Distribution Name       |  Ubuntu
Distribution Version    |  14.04, 16.04
Linux Kernel                 | 4.4 / 4.10
Architecture                 | amd64
ZFS Version                  | 0.6.5.6, 0.6.5.9, 0.7.0
SPL Version                  | 0.6.5.6, 0.6.5.9, 0.7.0

### Describe the problem you're observing

ZFS receive fails occassionaly when receiving many ZVOLs with snapshots/clones.

### Describe how to reproduce the problem

Either a recursive or a sequential receive of many datasets (=ZVOLs) sometimes fails with "dataset is busy".
The workaround is to either chmod -x /lib/udev/zvol_id, or to set zvol_inhibit_dev=1
Neither solution is nice in case one needs to use this copy (backup).

I trigger this almost any time I start a new full backup, sometimes even with incremental backup that contains new datasets. Iit happens on Ubuntu 14 with zfs-dkms as well as on two different machines with 16.04 (4.4 or 4.10 HWE kernel, even with ZFS 0.7.0). The source storage is fast (full flash) connected via 10G to the backup server. Backup server uses spindles.

What seems to happen is that once a ZVOL is received, zvol_id triggers and wants to open this ZVOL and create a symlink, however receive wants to do more work on that same dataset because it's also receiving a whole bunch of snapshots and clones of this dataset - this fails if zvol_id holds a lock.

### Proposed solutions

1) make zfs receive wait a little while for the lock before giving up
2) make zvol_id not lock the dataset

**3) allow user to configure device creation on a dataset level**
I have thousands of /dev/zd* devices on my backup storage and I don't even need them most of the time. What I need is to create a dev node for a particular dataset and then make it go away when I've used it. This setting could be for one singular dataset or a parent dataset and recursive.
I believe something like:
zfs set zvol_inhibit_dev=1 backup/zfs1.prodcluster/tank
Would be very useful

**Comment by gmelikov** [comment]
> 3) allow user to configure device creation on a dataset level

Please look at https://github.com/zfsonlinux/zfs/commit/cf8738d85374f51298a0872bcd58257bbb4fda6d , it's already implemented.

**Comment by zviratko** [comment]
I didn't know about that! :-)

Cool. Thanks!

**Comment by behlendorf** [comment]
Closing.  Already implemented in current release.

**Comment by zviratko** [comment]
I'm sorry, but that was only 1/3 :-)
Many folks will want ZVOLs exposed, yet will hit this issue as well. I still believe this race should be fixed so that receive works reliably in both cases.

**Comment by behlendorf** [comment]
Right.  To be clear the remaining issue here is occasional `zfs receive` failures.

> What seems to happen is that once a ZVOL is received, zvol_id triggers and wants to open this ZVOL and create a symlink, however receive wants to do more work on that same dataset because it's also receiving a whole bunch of snapshots and clones of this dataset - this fails if zvol_id holds a lock.

Another possible way to handle this would be to by default defer registering the new zvols with the kernel until the full receive of the dataset and all its snapshots complete.  @bprotopopov may have some thoughts about how best to handle this.

**Comment by zviratko** [comment]
Thank you.

**Comment by bprotopopov** [comment]
I would definitely take a look, but at this time, I am in between representative hw setups. I cannot reproduce this by recursively sending 100 zvols on my VM.

Is there a way to gather any details as to which lock/which code paths take it ?

**Comment by zviratko** [comment]
Try having lots of snapshots of the datasets as well. It seems to happen after the ZVOL itself is received, but before any subsequent snapshots (I don't know what gets generated into zfs send stream with -R and in what order). 
Now that I think of it, I don't think zfs receive should want to do more work on the same dataset, but one possibility is that (in my case) the datasets often have clones whose names are the same (in case that matters for ordering). So the stream will contain (I think) something like:

dataset@snap1
dataset-clone1@snap1
dataset-clone2@snap1

and so on. Each of these also has ~20 snapshots made by zfSnap (recursive).
